include/api/CAnomalyJob.h

```diff
@@ -144,12 +143,12 @@ class API_EXPORT CAnomalyJob : public CDataProcessor {
                 CFieldConfig& fieldConfig,
                 model::CAnomalyDetectorModelConfig& modelConfig,
                 core::CJsonOutputStreamWrapper& outputBuffer,
+                CPersistenceManager* persistenceManager,
```

Comment:
Definitely +1 to making this non-optional.

Another option, which would make the error impossible in production code, would be make this constructor protected and only expose a version taking the persistence manager by reference. We'd then need a `class CTestAnomalyJob : public api::CAnomalyJob` which sets it to null.

Since it is unlikely that we'll have much churn around this code, I'm happy for you to decide if this is worthwhile.

Reply:
Unfortunately I don't think the production code can pass the persistence manager by reference as there is an obscure option to never persist state:

https://github.com/elastic/elasticsearch/blob/0f2f00a20a6eae5b1fa9931f3b300746eb28e1c6/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectBuilder.java#L218-L219

However, good idea about having a derived class for tests.  I will make _all_ the arguments mandatory in production code and introduce a test class that defaults the same ones that were defaulted previously.  Then I don't need to change the argument order which will make the change cleaner.